src/cdr/resolver.go

```diff
@@ -0,0 +1,271 @@
+			rc, err = cacheRC(rc)
+		}
+	case *Ref_Compress:
+		rc, err = r.derefCompress(ctx, body.Compress, false)
+		if err != nil {
+			return nil, err
+		}
+		if cache && IsImmutable(ref) {
+			rc, err = cacheRC(rc)
+		}
+	case *Ref_Slice:
+		rc, err = r.derefSlice(ctx, body.Slice, cache)
+	case *Ref_Concat:
+		rc, err = r.derefConcat(ctx, body.Concat, cache)
+	default:
+		return nil, errors.Errorf("unsupported Ref variant %v: %T", ref.Body, ref.Body)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return rc, nil
+}
+
+func (r *Resolver) derefHTTP(ctx context.Context, ref *HTTP) (io.ReadCloser, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, ref.Url, nil)
+	if err != nil {
+		return nil, err
+	}
+	for k, v := range ref.Headers {
+		req.Header.Add(k, v)
+	}
+	// TODO: Check response code.
+	resp, err := r.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Body, nil
+}
+
+func (r *Resolver) derefContentHash(ctx context.Context, ref *ContentHash, cache bool) (io.ReadCloser, error) {
+	innerRc, err := r.deref(ctx, ref.Inner, cache)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: add a maximum limit here
+	data, err := io.ReadAll(innerRc)
+	if err != nil {
+		return nil, err
+	}
+	var h hash.Hash
+	switch ref.Algo {
+	case HashAlgo_BLAKE2b_256:
+		h, err = blake2b.New256(nil)
+		if err != nil {
+			panic(err)
+		}
+	default:
+		return nil, errors.Errorf("unrecognized hash algo %v", ref.Algo)
+	}
+	h.Write(data)
+	sum := h.Sum(nil)
+	if !bytes.Equal(sum, ref.Hash) {
+		return nil, errors.Errorf("content failed hash check HAVE: %x WANT: %x", sum, ref.Hash)
+	}
+	return io.NopCloser(bytes.NewReader(data)), nil
+}
+
+func (r *Resolver) derefCipher(ctx context.Context, ref *Cipher, cache bool) (io.ReadCloser, error) {
+	innerRc, err := r.deref(ctx, ref.Inner, cache)
+	if err != nil {
+		return nil, err
+	}
+	var rd io.Reader
+	switch ref.Algo {
+	case CipherAlgo_CHACHA20:
+		ciph, err := chacha20.NewUnauthenticatedCipher(ref.Key, ref.Nonce)
+		if err != nil {
+			return nil, err
+		}
+		rd = &cipher.StreamReader{R: innerRc, S: ciph}
+	default:
+		return nil, errors.Errorf("unrecognized cipher algo %v", ref.Algo)
+	}
+	rc := readCloser{r: rd, closes: []func() error{
+		innerRc.Close,
+	}}
+	return rc, nil
+}
+
+func (r *Resolver) derefCompress(ctx context.Context, ref *Compress, cache bool) (io.ReadCloser, error) {
+	innerRc, err := r.deref(ctx, ref.Inner, cache)
+	if err != nil {
+		return nil, err
+	}
+	switch ref.Algo {
+	case CompressAlgo_GZIP:
+		gr, err := gzip.NewReader(innerRc)
+		if err != nil {
+			return nil, err
+		}
+		return readCloser{r: gr, closes: []func() error{
+			innerRc.Close,
+			gr.Close,
+		}}, nil
+	default:
+		return nil, errors.Errorf("unrecognized compress algo %v", ref.Algo)
+	}
+}
+
+func (r *Resolver) derefSlice(ctx context.Context, ref *Slice, cache bool) (io.ReadCloser, error) {
+	innerRc, err := r.deref(ctx, ref.Inner, cache)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Is the purpose of the empty read here to confirm there are no errors reading back the data?

Reply:
It is to skip to the starting point.